search/request.go

```diff
@@ -0,0 +1,163 @@
+package search
+
+import (
+	"encoding/json"
+	"github.com/tigrisdata/tigris-client-go/driver"
+	"github.com/tigrisdata/tigris-client-go/fields"
+	"github.com/tigrisdata/tigris-client-go/filter"
+)
+
+// Request for search
+type Request struct {
+	Q            string
+	SearchFields []string
+	Filter       filter.Filter
+	Facet        *FacetQuery
+	ReadFields   fields.Read
+	Options      *RequestOptions
+}
+
+func NewRequestBuilder(q string) RequestBuilder {
+	return &requestBuilder{
+		q:            q,
+		searchFields: make(map[string]bool)}
+}
+
+type RequestBuilder interface {
+	WithSearchFields(fields ...string) RequestBuilder
+	WithFilter(filter.Filter) RequestBuilder
+	WithFacet(*FacetQuery) RequestBuilder
+	WithReadOptions(fields.Read) RequestBuilder
+	WithOptions(*RequestOptions) RequestBuilder
+	Build() *Request
+}
+
+type requestBuilder struct {
+	q            string
+	searchFields map[string]bool
+	filter       filter.Filter
+	facet        *FacetQuery
+	readFields   fields.Read
+	options      *RequestOptions
+}
+
+func (b *requestBuilder) WithSearchFields(fields ...string) RequestBuilder {
+	for _, f := range fields {
+		b.searchFields[f] = true
+	}
+	return b
+}
+
+func (b *requestBuilder) WithFilter(f filter.Filter) RequestBuilder {
+	b.filter = f
+	return b
+}
+
+func (b *requestBuilder) WithFacet(facet *FacetQuery) RequestBuilder {
+	b.facet = facet
+	return b
+}
+
+func (b *requestBuilder) WithReadOptions(read fields.Read) RequestBuilder {
+	b.readFields = read
+	return b
+}
+
+func (b *requestBuilder) WithOptions(options *RequestOptions) RequestBuilder {
+	b.options = options
+	return b
+}
+
+func (b *requestBuilder) Build() *Request {
```

Comment:
This builder looks more Java, any golang design pattern we should use instead? This is user facing construct.

Reply:
Take a look how update fields are implemented. 
`Update` is a builder itself, so user may or may not call `Build` explicitly.

Also set and unset defined on the package level, so no explicit object creation required.

In the case of search, Request would have all `With...` methods and they also will be defined on the package level, so the following possible:

```
coll1.Search(ctx, search.WithSearchFields(...).WithOptions(...))
```
The build is called implicitly in this case.
Explicit user workflow with object creation and Build call works as well.

Request would have unexported driver.SearchRequest embedded. Building would mean to compile driver.SearchRequest from high level Request.